src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -782,7 +783,14 @@ async Task RawOpen(SslMode sslMode, NpgsqlTimeout timeout, bool async, Cancellat
 
             IsSecure = false;
 
-            if ((sslMode is SslMode.Prefer && DataSource.TransportSecurityHandler.SupportEncryption) ||
+            if (Settings.SslNegotiation == SslNegotiation.Direct)
```

Comment:
Silly edge case, but what should happen if SslNegotiation is direct but SslMode is e.g. Disable? We should maybe throw on incompatible combinations here, and/or possibly check what libpq does.

Reply:
We should throw and that's already done in `NpgsqlConnectionStringBuilder.PostProcessAndValidate`

Reply:
I can copy the assert from there to here if you want...

Reply:
Ah right, missed the validation. Yeah, an assertion could be nice for clarity, if you want.